File: examples/python/set.py
# Copyright 2005-2015 ECMWF.
#
# This software is licensed under the terms of the Apache Licence Version 2.0
# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
#
# In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
# virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.

import traceback
import sys

from gribapi import *
from datetime import date

INPUT='../../data/regular_latlon_surface_constant.grib1'
OUTPUT='out.grib'
VERBOSE=1 # verbose error reporting

def example():
    fin = open(INPUT)
    fout = open(OUTPUT,'w')
    gid = grib_new_from_file(fin)

    dt = date.today()
    today = "%d%02d%02d" % (dt.year,dt.month,dt.day)
    grib_set(gid,'dataDate',int(today))
    grib_set(gid,'centre',80)

    centreIntVal = grib_get(gid,'centre',int)
    centreStrVal = grib_get(gid,'centre',str)
    dateStrVal   = grib_get(gid,'dataDate',str)
    assert(centreIntVal == 80)
    assert(centreStrVal == 'cnmc')
    assert(dateStrVal   == today)
    print 'get centre as an integer - centre = %d' % centreIntVal
    print 'get centre as a string - centre = %s' % centreStrVal
    print 'get date as a string - date = %s' % dateStrVal
    
    # Now do the same but using set_key_vals, setting keys all at once
    grib_set_key_vals(gid, 'level=1,centre=98')  # with a String
    assert(grib_get(gid,'centre',str) == 'ecmf')
    assert(grib_get(gid,'level',int)  == 1)
    
    grib_set_key_vals(gid, ['level=2', 'centre=kwbc']) # with a Tuple
    assert(grib_get(gid,'centre',int) == 7)
    assert(grib_get(gid,'level',int)  == 2)
    
    grib_set_key_vals(gid, {'level': 3, 'centre': 84}) # with a Dictionary
    assert(grib_get(gid,'centre',str) == 'lfpw')
    assert(grib_get(gid,'level',int)  == 3)
    
    grib_gts_header(True)
    grib_gts_header(False)

    grib_write(gid,fout)
    grib_release(gid)
    fin.close()
    fout.close()

def main():
    try:
        example()
    except GribInternalError,err:
        if VERBOSE:
            traceback.print_exc(file=sys.stderr)
        else:
            print >>sys.stderr,err.msg

        return 1

if __name__ == "__main__":
    sys.exit(main())

File: examples/F90/set.f90
! Copyright 2005-2014 ECMWF.
!
! This software is licensed under the terms of the Apache Licence Version 2.0
! which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
! 
! In applying this licence, ECMWF does not waive the privileges and immunities granted to it by
! virtue of its status as an intergovernmental organisation nor does it submit to any jurisdiction.
!
!
!
!  Description: how to set key values.
!
!
!
program set
  use grib_api
  implicit none
  integer(kind = 4)    :: centre, date1
  integer              :: infile,outfile
  integer              :: igrib


  centre = 80
  call current_date(date1)
  call grib_open_file(infile, &
       '../../data/regular_latlon_surface_constant.grib1','r')

  call grib_open_file(outfile, &
       'out.grib1','w')

  !     a new grib message is loaded from file
  !     igrib is the grib id to be used in subsequent calls
  call grib_new_from_file(infile,igrib)

  call grib_set(igrib,'dataDate',date1)
  !     set centre as a integer */
  call grib_set(igrib,'centre',centre)

! check if it is correct in the actual GRIB message

  call check_settings(igrib)

  !     write modified message to a file
  call grib_write(igrib,outfile)

  call grib_release(igrib)

  call grib_close_file(infile)

  call grib_close_file(outfile)

contains

!======================================
subroutine current_date(date1)
integer, intent(out) :: date1

integer              :: val_date(8)
call date_and_time ( values = val_date)

date1 = val_date(1)* 10000 + val_date(2)*100 + val_date(3) 
end subroutine current_date
!======================================
subroutine check_settings(gribid)
  implicit none
  integer, intent(in) :: gribid
  
  integer(kind = 4)    :: int_value
  character(len = 10)  :: string_value

  !     get centre as a integer
  call grib_get(gribid,'centre',int_value)
  write(*,*) "get centre as a integer - centre = ",int_value
  
  !     get centre as a string
  call grib_get(gribid,'centre',string_value)
  write(*,*) "get centre as a string  - centre = ",string_value
  
  !     get date as a string
  call grib_get(gribid,'dataDate',string_value)
  write(*,*) "get date as a string    - date = ",string_value
  
end subroutine check_settings
end program set
